src/Microsoft.VisualStudio.LanguageServices.Razor/VsSolutionUpdatesProjectSnapshotChangeTrigger.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.LanguageServices.Razor.Editor;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    [Export(typeof(ProjectSnapshotChangeTrigger))]
+    internal class VsSolutionUpdatesProjectSnapshotChangeTrigger : ProjectSnapshotChangeTrigger
+    {
+        private readonly IServiceProvider _services;
+        private readonly TextBufferProjectService _projectService;
+
+        [ImportingConstructor]
+        public VsSolutionUpdatesProjectSnapshotChangeTrigger(
+            [Import(typeof(SVsServiceProvider))] IServiceProvider services,
+            TextBufferProjectService projectService)
+        {
+            _services = services;
+            _projectService = projectService;
+        }
+
+        public override void Initialize(ProjectSnapshotManagerBase projectManager)
+        {
+            var eventSink = new SolutionUpdateEventSink(projectManager, _projectService);
+
+            // Attach the event sink to solution update events.
+            var solutionBuildManager = _services.GetService(typeof(SVsSolutionBuildManager)) as IVsSolutionBuildManager;
+            var result = solutionBuildManager.AdviseUpdateSolutionEvents(eventSink, out var cookie);
+            Debug.Assert(result == VSConstants.S_OK);
+        }
+
+        // Internal for testing.
+        internal class SolutionUpdateEventSink : IVsUpdateSolutionEvents2
```

Comment:
Consider just implementing these events on the class itself. Roslyn has a nice pattern for this. When they have to implement an API like this, they put that in a partial class in another file and then call protected/internal methods with the relevant data.